pkg/conditions/infrastructureready/handler.go

```diff
@@ -0,0 +1,128 @@
+package infrastructureready
+
+import (
+	"context"
+
+	"github.com/giantswarm/conditions/pkg/conditions"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	capi "sigs.k8s.io/cluster-api/api/v1alpha3"
+	capiexternal "sigs.k8s.io/cluster-api/controllers/external"
+	capiexp "sigs.k8s.io/cluster-api/exp/api/v1alpha3"
+	capiconditions "sigs.k8s.io/cluster-api/util/conditions"
+	ctrl "sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/giantswarm/conditions-handler/pkg/internal"
+	"github.com/giantswarm/conditions-handler/pkg/key"
+)
+
+type HandlerConfig struct {
+	CtrlClient ctrl.Client
+	Logger     micrologger.Logger
+
+	Name         string
+	UpdateStatus bool
+}
+
+type Handler struct {
+	ctrlClient      ctrl.Client
+	internalHandler *internal.Handler
+	logger          micrologger.Logger
+	name            string
+}
+
+func NewHandler(config HandlerConfig) (*Handler, error) {
+	h := &Handler{
+		ctrlClient: config.CtrlClient,
+		logger:     config.Logger,
+		name:       config.Name,
+	}
+
+	internalHandlerConfig := internal.HandlerConfig{
+		CtrlClient:        config.CtrlClient,
+		Logger:            config.Logger,
+		UpdateStatus:      config.UpdateStatus,
+		ConditionType:     capi.InfrastructureReadyCondition,
+		EnsureCreatedFunc: h.ensureCreated,
+	}
+
+	internalHandler, err := internal.NewHandler(internalHandlerConfig)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+	h.internalHandler = internalHandler
+
+	return h, nil
+}
+
+func (h *Handler) EnsureCreated(ctx context.Context, object interface{}) error {
+	obj, err := key.ToObjectWithConditions(object)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	return h.internalHandler.EnsureCreated(ctx, obj)
+}
+
+func (h *Handler) EnsureDeleted(_ context.Context, _ interface{}) error {
+	return nil
+}
+
+func (h *Handler) Name() string {
+	return h.name
+}
+
+func (h *Handler) ensureCreated(ctx context.Context, object conditions.Object) error {
+	obj, err := toObjectWithInfrastructure(object)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	infrastructureObject, err := h.getInfrastructureObject(ctx, obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	update(obj, infrastructureObject)
+	return nil
+}
+
+func (h *Handler) getInfrastructureObject(ctx context.Context, object objectWithInfrastructureRef) (capiconditions.Getter, error) {
+	infrastructureRef := object.GetInfrastructureRef()
+	if infrastructureRef == nil {
+		// Infrastructure object is not set
+		return nil, nil
+	}
+
+	infrastructureObject, err := capiexternal.Get(ctx, h.ctrlClient, object.GetInfrastructureRef(), object.GetNamespace())
+	if apierrors.IsNotFound(err) {
```

Comment:
Based on tests for `ControlPlaneReady` that is using `ControlPlaneRef` to fetch control plane object, when the object is not found this `apierrors.IsNotFound(err)` call will probably not get that, as CAPI is wrapping the error.

I'll refactor it like it's done for `ControlPlaneReady` (which is covered with tests more).